### DISCUSSION TITLE
[question] How to parse repeated, line wise, blocks without having to duplicate the block spec for the last line with eof instead crlf

### DISCUSSION THREAD

**Discussion by ridilculous** [tree-sitter/tree-sitter#2958] [Q&A] [answered]
I'm working on a line wise parser and cannot figure how to create a line rule which can be used for the first _n-1_ lines of the document which end with `[\r\n]+` and the last line which "ends" with `eof`.

This is the grammar i'm working with (i tried a lot but this is the only working spec so far and for the last line, the `$.block` rule has to be duplicated into the `$.source_file` rule:

```javascript
module.exports = grammar({
  name: 'test',

  conflicts: $ => [[$.block]],
  extras: $ => [],
  externals: $ => [$.eof],

  rules: {
    source_file: $ => seq(
      repeat($.block),
      seq(
        repeat($.n),
        $.eof
    )),

    block: $ => seq(
      repeat($.n),
      /[\r\n]+/,
    ),

    n: $ => /[Nn]/,
  }
});
```

The external scanner for the eof is very simple:

```c++
struct Scanner {
  bool scan(TSLexer *lexer, const bool *valid_symbols) {
    if (lexer->eof(lexer)) {
      lexer->advance(lexer, false);
      lexer->mark_end(lexer);
      lexer->result_symbol = 0;
      return true;
    }
    return false;
}};
```

It feels like i'm overlooking something really simple and obvious but after many attempts i'm still failing, either with 

> Tree-sitter does not support syntactic rules that match the empty string unless they are used only as the grammar's start rule.

or

a infinite-loop when i try to do things like
`choice(/[\r\n]+/, $.eof),`
at the end of the block.

I even tried advancing the lexer twice on eof.

Is there a simpler way without the duplication?

**Comment by amaanq** [comment, marked as answer]
If the null character isn't allowed in the language at all, you can just add that to block which would match the eof case as well, otherwise you have to pick using the scanner or duplicating most of the rule
